src/NuGet.Indexing/Extraction/DocumentCreator.cs

```diff
@@ -261,6 +262,61 @@ private static void AddRequiresLicenseAcceptance(Document document, IDictionary<
             }
         }
 
+        private static void AddPackageTypes(Document document, IDictionary<string, string> package)
+        {
+            string flattenedPackageTypes;
+            StringBuilder searchIndex = new StringBuilder();
+            if (package.TryGetValue("flattenedPackageTypes", out flattenedPackageTypes))
+            {
+                if (!string.IsNullOrWhiteSpace(flattenedPackageTypes))
+                {
+                    AddField(document, "FlattenedPackageTypes", flattenedPackageTypes, Field.Index.NOT_ANALYZED);
+
+                    using (var textWriter = new StringWriter())
+                    {
+                        using (var jsonWriter = new JsonTextWriter(textWriter))
+                        {
+                            jsonWriter.WriteStartArray();
+                            foreach (var packageType in flattenedPackageTypes.Split('|'))
+                            {
+                                string[] fields = packageType.Split(':');
+                                if (fields.Length > 0)
+                                {
+                                    jsonWriter.WriteStartObject();
+                                    jsonWriter.WritePropertyName("Type");
+                                    jsonWriter.WriteValue(fields[0]);
+                                    searchIndex.Append(fields[0]);
+                                    searchIndex.Append(" ");
+                                    if (fields.Length > 1)
+                                    {
+                                        jsonWriter.WritePropertyName("Version");
+                                        jsonWriter.WriteValue(fields[1]);
+                                    }
+
+                                    jsonWriter.WriteEndObject();
+                                }
+                            }
+
+                            jsonWriter.WriteEndArray();
+                            jsonWriter.Flush();
+                            textWriter.Flush();
+                            var packageTypes = textWriter.ToString();
+
+                            AddField(document, "packageTypes", packageTypes, Field.Index.NOT_ANALYZED);
+                        }
+                    }
+                }
+            }
+
+            // remove the trailing space from the index before saving
+            if (searchIndex.Length > 0)
+            {
+                searchIndex.Length--;
+            }
+
+            AddField(document, "PackageTypesIndex", searchIndex.ToString(), Field.Index.ANALYZED);
```

Comment:
In addition, analyzing the JSON data is a lot of work for Lucene, both during index and during query. Might want to make this field contain just the values instead of full-blwon JSON.